### PR TITLE
Send new user message to the admin_logs channel

### LINKED ISSUE #32: Add new member reporting to the admin_logs channel
## Description
The admins would like to receive messages when a new user joins the workspace along with their email. If at all possible, we'd also like to pin these messages automatically so they can be viewed later. This message should include the user's id, display name, full name, and email. We should also add an @ with their user id so if their display name changes, the mention will automatically update.

## Open questions
- Is there a pin limit?
- Is there anything else we'd like to include in the message?

## Acceptance Criteria
- [ ] When a new user joins the workspace, send a message to the admin channel with the user's name, id, email, and mention
- [ ] Pin this message to the channel so it doesn't get eaten by the message limit

## Resources
Will need to add the [`user:read` scope](https://api.slack.com/scopes/users:read) and the [`pins:write` scope](https://api.slack.com/scopes/pins:write)
[`team_join` event "docs"](https://api.slack.com/events/team_join)
Docs seem sparse but it looks like we should get the following
```json
{
    "type": "team_join",
    "user": {
        "id": "W012A3CDE",
        "team_id": "T012AB3C4",
        "name": "spengler",
        "deleted": false,
        "color": "9f69e7",
        "real_name": "Egon Spengler",
        "tz": "America/Los_Angeles",
        "tz_label": "Pacific Daylight Time",
        "tz_offset": -25200,
        "profile": {
            "avatar_hash": "ge3b51ca72de",
            "status_text": "Print is dead",
            "status_emoji": ":books:",
            "real_name": "Egon Spengler",
            "display_name": "spengler",
            "real_name_normalized": "Egon Spengler",
            "display_name_normalized": "spengler",
            "email": "spengler@ghostbusters.example.com",
            "image_original": "https://.../avatar/e3b51ca72dee4ef87916ae2b9240df50.jpg",
            "image_24": "https://.../avatar/e3b51ca72dee4ef87916ae2b9240df50.jpg",
            "image_32": "https://.../avatar/e3b51ca72dee4ef87916ae2b9240df50.jpg",
            "image_48": "https://.../avatar/e3b51ca72dee4ef87916ae2b9240df50.jpg",
            "image_72": "https://.../avatar/e3b51ca72dee4ef87916ae2b9240df50.jpg",
            "image_192": "https://.../avatar/e3b51ca72dee4ef87916ae2b9240df50.jpg",
            "image_512": "https://.../avatar/e3b51ca72dee4ef87916ae2b9240df50.jpg",
            "team": "T012AB3C4"
        },
        "is_admin": true,
        "is_owner": false,
        "is_primary_owner": false,
        "is_restricted": false,
        "is_ultra_restricted": false,
        "is_bot": false,
        "updated": 1502138686,
        "is_app_user": false,
        "has_2fa": false
    }
}
```
Post message response should include all we need to pin the message ([docs](https://api.slack.com/methods/chat.postMessage#examples))
[`pins_add` method docs](https://api.slack.com/methods/pins.add)

### PR DESCRIPTION
**Pull request by hschallh**
Closes #32 

## Description
Adds a new subscription to team_join events. Crafts a message with some basic user info and sends it to the admin_logs channel. Automod then automatically pins this message to the channel.

### Screenshots
Screen Shot 2021-12-18 at 3 16 18 PM

## AC
- [ ] When a new user joins the workspace, send a message to the admin channel with the user's name, id, email, and mention
- [ ] Pin this message to the channel so it doesn't get eaten by the message limit

### Initial setup
1. Clone this repository and check out this branch
    ```bash
    git clone git@github.com:OSU-CS/automod.git
    git checkout hts/new-user-message
    ```
2. Install python3.7 and pip
3. Create base venv (virtual environment)
    ```bash
    pip install virtualenv
    virtualenv -p `which python3.7` venv
    ```
4. Activate venv
    ```bash
    source venv/bin/activate
    ```
5. Rebuild venv with dependencies
    ```bash
    pip install -Ur requirements.txt
    ```
6. Run app (after activating venv). (Ask hunter)
    ```bash
    SLACK_BOT_TOKEN=<intergration_oauth_key>
    export SLACK_BOT_TOKEN
    python app.py
    ```
### Testing functionality
1. Ask Hunter for an invite to the integration testing slack workspace and an invite to the `admin_logs` channel
2. Sign up for this workspace
3. Confirm message was sent with the correct information
4. Confirm this message was pinned
5. Change your display name
6. Confirm it was updated in the mention of the message

### CODE REVIEW

File: app.py [hunk 1]
```diff
@@ -37,6 +38,18 @@ def new_channel_callback(**payload) -> None:
...
+@slack.RTMClient.run_on(event='team_join')
+def new_user_callback(**payload) -> None:
+    """Catch team_join event.
+
+    Triggered when a user has joined the workspace.
+    """
+    web_client = payload['web_client']
+    new_user = payload['data']['user']
+
+    send_new_user_message(web_client, 'admin_logs', new_user)
```
Comment (sarahmaas) [inline app.py:50]: At some point do we want to start abstracting this out into at least constants somewhere if not an environment configuration?

File: app.py [hunk 2]
```diff
@@ -84,6 +97,21 @@ def send_new_channel_message(web_client: slack.WebClient, report_channel: str, n
...
+    :param report_channel: The channel to send the message to
+    :param new_user: The new slack user object
+    """
+    new_user_message = NewUserMessage(report_channel, new_user)
+    message = new_user_message.get_message_payload()
+
+    response = web_client.chat_postMessage(**message)
+
+    # Pin this message to the channel
+    web_client.pins_add(channel=response['channel'], timestamp=response['ts'])
```
Comment (sarahmaas) [inline app.py:113]: How do we handle if pins are full?

Comment (hschallh) [reply app.py:113]: cry?

Comment (hschallh) [reply app.py:113]: We could check for a "too_many_pins" error, get all the pins in the channel, remove the last one, and retry, but looking at the docs, bots can't yet get the list of pins of a channel???

File: app.py (hunk 1 above)
Comment (hschallh) [reply app.py:50]: Absolutely. Let's write a ticket for this.

File: app.py (hunk 2 above)
Comment (sarahmaas) [reply app.py:113]: Hmm.... For now maybe just send a message that tags @channel when it happens so we can manually handle? Not ideal but it's something

Comment (hschallh) [reply app.py:113]: I'll merge this now and we can just manually keep an eye on the number of pins, removing the old ones. I'll keep checking the docs to see if they add support for app bots to check in on pins.

Comment (hschallh) [reply app.py:113]: https://github.com/OSU-CS/automod/issues/34

**Pull request merged**